### ISSUE TITLE
Configuration database of art/ROOT file missing last job configuration

### ISSUE THREAD

**Issue by PetrilloAtWork** [art-framework-suite/art-root-io#16]
## Bug description

The problem is that processing an ICARUS (data) file with a standard reconstruction configuration ("stage0"), the configuration database of the ROOT output file seems not to list that configuration and process.
`file_info_dumper` correctly reports two processes having happened, but a configuration dump shows only the process in the input file (`DAQEVB04`) and not the one of the job that created the file (`stage0`).

Even stranger, if I cut out either an unused service or a specific module, the `stage0` configuration is instead correctly written. In the past I did not notice this effect, so it may even be due to some changes in art. Or not.

## To Reproduce

In `/exp/icarus/data/users/petrillo/for/art`, accessible from an ICARUS GPVM, there is an input ROOT file and three configurations to show the issue (genially named "good1", "good2" and "broken"). Setup instructions are written on the top of each FHiCL file, and that's just:
    
    source /cvmfs/icarus.opensciencegrid.org/products/icarus/setup_icarus.sh
    setup icaruscode v09_82_01 -q e26:prof
    
I checked the outcome of each using `config_dumper -P -f stage0 test.root`.

### Whether this is a high-priority issue

The priority depends on the nature of the issue: if it's a real, general problem, it should be addressed before ICARUS starts taking "Run3" data (1-2 weeks). If it's only my problem, then the solution schedule is more relaxed.

**Users can't add labels to the issues anyway.**

**Comment by PetrilloAtWork** [comment]
It appears that the issue has hit our standard sample processing too.
Because of that I can say that the issue was not observed in the previous processing campaigns, using `icaruscode` `v09_78_04_01` or similar, but are present with the current ones (e.g. `v09_82_02_01`). Both sport the same version of `art_root_io`; but the former is with GCC 9 (`e20`) while the latter is with GCC 12 (`e26`). We don't have the opposite builds, nor Clang ones, to further test this (and it could be a red herring).

**Comment by knoepfel** [comment]
@PetrilloAtWork, the on-disk database for just the input file has some **huge** configuration values from the DAQ.  Looks like something dumped a few million characters into the value of one configuration parameter?  Is it possible the DAQ has changed recently?  I wouldn't be surprised if SQLite3 just "gave up" because it was too much.  But I'd have to do more digging.  Incidentally, art does record the `stage0` process in its non-configuration metadata:

```console
$ file_info_dumper --process-history test.root 
==============================
File: test.root
Warning in <TClass::Init>: no dictionary for class sim::IDE is available

 Chronological list of process names for processes that
 produced this file.

    1. DAQEVB04
    2. stage0
```

where `test.root` was created with the "broken" configuration.

**Comment by PetrilloAtWork** [comment]
I confirm that DAQ dumps readout/DAQ configuration into that database.
The configuration is typically FHiCL, flattened into a string (`ParameterSet::to_string()` I guess). This has not changed.
Also note that the issue is seen also processing raw data produced on January 2023, and that the past processing with older software did not show in the past the issue.
I confirm that I did see the two processes listed since the beginning. But the database seems to be incomplete/inaccessible/troubled. In addition, the database in the raw file 
(that is the file that comes directly from DAQ) is fully accessible too.

**Comment by knoepfel** [comment]
I have standalone code intended to reproduce this problem, and so far I am unable to reproduce it.  I am running valgrind now with the hopes that it will tell me something.

**Comment by knoepfel** [comment]
@PetrilloAtWork, although there is an issue we need to understand here, can you tell us what you need `config_dumper` for?  IOW, I'm trying to anticipate if in the meantime we need an alternative solution to what you're trying to solve by using `config_dumper`.

**Comment by PetrilloAtWork** [comment]
> I have standalone code intended to reproduce this problem, and so far I am unable to reproduce it. I am running valgrind now with the hopes that it will tell me something.

Are you able to reproduce the error from the ICARUS file though?

**Comment by knoepfel** [comment]
Yes, I can reproduce the problem.  `valgrind` yielded nothing.  Still working on it.

**Comment by PetrilloAtWork** [comment]
I am not too surprised it's hard to reproduce -- considering that removing an unrelated service makes the problem disappear...

As for the need, there is no need for `config_dumper`. I quoted that to prove that the issue is not triggered specifically by ICARUS code.

The code being broken needs to consult the DAQ configuration. The name of the DAQ process is not unique, and `art::Event` interface requires a process name to get its configuration (no "global" configuration that I am aware of). So our algorithm looks for patterns on each of the configurations, looping through all the available process names. It turns out that the DAQ process name appears to be available, so in principle we _could_ work around this in this case (provided that the defective pattern is consistent) by catching and ignoring the exception.
But I would like not to produce data run files that are defective, and especially not the raw data ones from DAQ, since we'll have to live with them forever.

To be clear, it's not even clear to me that the issue originates from the raw files, and in fact I tend to think it does not. But I would like the issue to be understood, to be sure.

**Comment by knoepfel** [comment]
@PetrilloAtWork, the problem is understood, and it is an error deep within `fhiclcpp`.  I will discuss the issue with the SciSoft team early next week and we will formulate a path forward.  This will mean a new release of art, etc.
